### PR TITLE
[GLUTEN-4116][VL] Support other object store server scheme in native write

### PR DESCRIPTION
**Pull request by dcoliversun**
## What changes were proposed in this pull request?

This PR aims to support other object store server scheme in native write.

ref: https://facebookincubator.github.io/velox/develop/connectors.html#storage-adapters

(Fixes: \#4116)

## How was this patch tested?

manual tests

### CODE REVIEW

**Comment by dcoliversun** [comment]
test sql:
```sql
create database if not exists db location 'oss://<bucket>/<db_path>/';
create table db.t1(a INT) STORED AS PARQUET PARTITIONED BY (b INT, c INT) LOCATION 'oss://<bucket>/<db_path>/<tbl_path>';
insert into table db.t1 partition(b=1, c=10) SELECT 1 AS a ;
```

Got log
```plain
23/12/20 04:32:17 WARN HiveConf: HiveConf of name hive.stats.jdbc.timeout does not exist
23/12/20 04:32:17 WARN HiveConf: HiveConf of name hive.stats.retries.wait does not exist
23/12/20 04:32:21 WARN ObjectStore: Version information not found in metastore. hive.metastore.schema.verification is not enabled so recording the schema version 2.3.0
23/12/20 04:32:21 WARN ObjectStore: setMetaStoreSchemaVersion called but recording version is disabled: version = 2.3.0, comment = Set by MetaStore UNKNOWN@172.17.0.2
23/12/20 04:32:21 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
23/12/20 04:32:26 WARN SessionState: METASTORE_FILTER_HOOK will be ignored, since hive.security.authorization.manager is set to instance of HiveAuthorizerFactory.
23/12/20 04:32:26 WARN HiveConf: HiveConf of name hive.internal.ss.authz.settings.applied.marker does not exist
23/12/20 04:32:26 WARN HiveConf: HiveConf of name hive.stats.jdbc.timeout does not exist
23/12/20 04:32:26 WARN HiveConf: HiveConf of name hive.stats.retries.wait does not exist
E1220 04:32:45.267411 20728 Exceptions.h:69] Line: ../../velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:242, Function:Impl, Expression:  Failed to create S3 bucket due to: 'Access denied'. Path:'s3://henghzhen-test-hangzhou/', SDK Error Type:15, HTTP Status Code:403, S3 Service:'AmazonS3', Message:'Access Denied', RequestID:'GAYTYC04QSSVBFYQ', Source: RUNTIME, ErrorCode: INVALID_STATE
23/12/20 04:32:45 ERROR Utils: Aborting task
java.lang.RuntimeException: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Failed to create S3 bucket due to: 'Access denied'. Path:'s3://henghzhen-test-hangzhou/', SDK Error Type:15, HTTP Status Code:403, S3 Service:'AmazonS3', Message:'Access Denied', RequestID:'GAYTYC04QSSVBFYQ'
Retriable: False
Function: Impl
File: ../../velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Line: 242
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 3  _ZN8facebook5velox11filesystems11S3WriteFile4ImplC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPN3Aws2S38S3ClientEPNS0_6memory10MemoryPoolE
# 4  _ZN8facebook5velox11filesystems11S3WriteFileC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPN3Aws2S38S3ClientEPNS0_6memory10MemoryPoolE
# 5  _ZN8facebook5velox11filesystems12S3FileSystem16openFileForWriteESt17basic_string_viewIcSt11char_traitsIcEERKNS1_11FileOptionsE
# 6  _ZN6gluten22VeloxParquetDatasource4initERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES7_St4hashIS7_ESt8equal_toIS7_ESaISt4pairIKS7_S7_EEE
# 7  Java_io_glutenproject_datasource_DatasourceJniWrapper_nativeInitDatasource
# 8  0x00007f64f8e0e427

	at io.glutenproject.datasource.DatasourceJniWrapper.nativeInitDatasource(Native Method)
	at io.glutenproject.datasource.DatasourceJniWrapper.nativeInitDatasource(DatasourceJniWrapper.java:49)
	at org.apache.spark.sql.execution.datasources.velox.VeloxFormatWriterInjects.createOutputWriter(VeloxFormatWriterInjects.scala:70)
	at org.apache.spark.sql.execution.datasources.velox.VeloxFormatWriterInjects.createOutputWriter$(VeloxFormatWriterInjects.scala:43)
	at org.apache.spark.sql.execution.datasources.velox.VeloxParquetWriterInjects.createOutputWriter(VeloxParquetWriterInjects.scala:26)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anon$1.newInstance(ParquetFileFormat.scala:108)
	at org.apache.spark.sql.execution.datasources.BaseDynamicPartitionDataWriter.renewCurrentWriter(FileFormatDataWriter.scala:311)
	at org.apache.spark.sql.execution.datasources.DynamicPartitionDataSingleWriter.beforeWrite(FileFormatDataWriter.scala:398)
	at org.apache.spark.sql.execution.datasources.DynamicPartitionDataSingleWriter.write(FileFormatDataWriter.scala:418)
	at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.writeWithMetrics(FileFormatDataWriter.scala:96)
	at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.writeWithIterator(FileFormatDataWriter.scala:103)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$executeTask$1(FileFormatWriter.scala:414)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1525)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.executeTask(FileFormatWriter.scala:423)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$write$17(FileFormatWriter.scala:331)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Please ignore error caused by no configuration pushed to s3 sdk in velox and it will be fixed soon. We can see s3 sdk in velox can support oss file in native write.

File: cpp/velox/operators/writer/VeloxParquetDatasource.h [hunk 1]
```diff
@@ -69,6 +69,19 @@ class VeloxParquetDatasource final : public Datasource {
     return schema_;
   }
 
+  bool isSupportedS3SdkPath(const std::string& filePath_) {
+    // support scheme
+    const std::array<const char*, 4> supported_schemes = {"s3a:", "oss:", "cos:", "cosn:"};
```
Comment (FelixYBW) [inline cpp/velox/operators/writer/VeloxParquetDatasource.h:74]: do we support s3:// ? I remember we can support with a config, @yma11 , this one?
.set('spark.hadoop.fs.s3.impl','org.apache.hadoop.fs.s3a.S3AFileSystem')

Comment (yma11) [reply cpp/velox/operators/writer/VeloxParquetDatasource.h:74]: I remember it doesn't work if the path started with `s3://`, we don't have changed the conf as you mentioned. As mentioned [here](https://hadoop.apache.org/docs/r3.1.1/hadoop-aws/tools/hadoop-aws/index.html#Other_S3_Connectors), `s3a` should be the one we support.

Comment (FelixYBW) [reply cpp/velox/operators/writer/VeloxParquetDatasource.h:74]: Just tried, with the config, we can read s3:// files. Looks it's a spark feature

Comment (dcoliversun) [reply cpp/velox/operators/writer/VeloxParquetDatasource.h]: thanks all, `s3` scheme added

**Review by FelixYBW — APPROVED** [review]

**Comment by dcoliversun** [comment]
@FelixYBW Github aciton fails with 137, I believe it is ok. This pr is ready to merge :)

**Pull request merged**